server/paymentProviders/braintree/gateway.ts

```diff
@@ -0,0 +1,189 @@
+    const nonce = paymentMethod.token;
+    if (customer) {
+      customer = await updateCustomer(gateway, customerId, nonce);
+    } else {
+      customer = await createBraintreeCustomer(gateway, fromCollective, nonce);
+    }
+
+    const token = customer.paymentMethods[0].token;
+    order.paymentMethod = await paymentMethod.update({
+      token,
+      data: { ...paymentMethod.data, isNonce: false, customerId: customer.id },
+    });
+  }
+
+  return customer;
+};
+
+export const callTransactionSale = async (
+  gateway: braintree.BraintreeGateway,
+  order: typeof models.Order,
+  isClientInitiated: boolean,
+): Promise<braintree.Transaction> => {
+  const response = await gateway.transaction.sale({
+    amount: (order.totalAmount / 100).toString(),
+    paymentMethodToken: order.paymentMethod.token,
+    deviceData: isClientInitiated ? order.paymentMethod.data?.deviceData : undefined,
+    transactionSource: order.interval ? 'recurring_first' : undefined,
+    customFields: {
+      collective: order.collective.slug,
+      collectiveId: order.collective.id,
+      order: order.id,
+    },
+    options: {
+      submitForSettlement: true,
+    },
+  });
+
+  if (!response.success) {
+    // TODO Handle errors
+    throw new Error(response.message);
+  }
+
+  return response.transaction;
+};
+
+export const callCreateSubscription = async (
+  gateway: braintree.BraintreeGateway,
+  order: typeof models.Order,
+): Promise<braintree.Transaction> => {
+  const result = await gateway.subscription.create({
+    paymentMethodToken: order.paymentMethod.token,
+    planId: order.interval === INTERVALS.MONTH ? MONTHLY_PLAN_ID : YEARLY_PLAN_ID,
+    neverExpires: true,
+    price: (order.totalAmount / 100).toString(),
+    options: {
+      startImmediately: true,
+      paypal: {
+        description: order.description,
+      },
+    },
+  });
+
+  if (!result.success) {
+    logger.error(`Failed to create subscription on Braintree: ${result.message}`);
+    logger.error(result);
+    throw new Error(`Failed to create subscription for order #${order.id}`);
+  } else {
+    const subscriptionId = result.subscription.id;
+    const braintreeData = { ...(order.data?.braintree || {}), subscriptionId };
+    await order.update({ data: { ...order.data, braintree: braintreeData } });
+    return result.subscription.transactions[0];
+  }
+};
+
+export const generateBraintreeTokenForClient = async (
+  gateway: braintree.BraintreeGateway,
+  fromCollective: typeof models.Collective | null = null,
+): Promise<string> => {
+  const customerId = fromCollective && (await getCustomerIdFromCollective(fromCollective));
+  const response = await gateway.clientToken.generate({ customerId });
+  if (!response.success) {
+    logger.error(`Failed to generate Braintree token for customer ${customerId || ''}: ${response.message}`);
+    throw new Error('Failed to generate Braintree token');
+  } else {
+    return response.clientToken;
+  }
+};
```

Comment:
Naming that `braintree.js` is not a bad idea. I took the inspiration for "gateway" from PayPal's implementation, where the file dealing with the requests is named gateway too.

I like having `helpers` as a separate file because it contains functions that do not interact with the gateway (aka. the Braintree API). In `gateway.js`, all functions either return or use a `braintree.BraintreeGateway`. `helpers` are things related to Braintree but that do not interact with their API.

That being said, I'm not against merging. They both have very similar purposes and the code is short enough.

Reply:
No strong opinions either.
I think these functions are already dealing with some of our logic so I'm fine with the current implementation.
In my head, `lib/` files should deal with generic and their service types only, and functions that are interacting with our data structures should live in `paymentProvider/`.